### PR TITLE
Migrate apiserver from utils/trace to component-base/tracing

### PR DESCRIPTION
**Pull request by dashpole**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Part of https://github.com/kubernetes/kubernetes/issues/113170

Depends on https://github.com/kubernetes/kubernetes/pull/113183.  See that PR for the proposed changes to utils/traces interfaces.

After this change, APIServerTracing traces will include spans from calls to tracing.Start(), and span events from calls to span.AddEvent().  **Log-based "tracing" will continue to function exactly the same, with the exception a change to the wording of `Step()` calls that included an error.**

If the APIServerTracing feature is disabled, the OpenTelemetry portion of these calls are no-ops.

#### Special notes for your reviewer:

Naming differences between utils/trace and component-base/tracing:

* `utiltrace.New()`, and `utiltrace.Nest()` are replaced by `tracing.Start()`
    * the thing returned by this call is now called `span` instead of `trace`, which is more accurate
* `trace.Step()` is replaced by `span.AddEvent()`
* `trace.LogIfLong()` is replaced by `span.End()`
* `utiltrace.Field` is replaced by `attribute.KeyValue`

This is intended to simply be a refactor, and shouldn't significantly change behavior when using the logging trace functionality.  The only behavioral change is that `trace.Step` calls which specify an error have been split into a "succeeded" step without an error, and a "failed" step with an error.  This is because err.Error() panics if err is nil.

In cases where a utiltrace.Trace was stored on an object, or passed via function, i've replaced those with Context-based passing of the current active span.

The component-base/tracing library gracefully handles invoking functions on spans retrieved from context.  It is no longer necessary to check for nil before calling AddEvent or LogIfLong.

#### Does this PR introduce a user-facing change?
```release-note
API Server Tracing now includes a variety of new spans and span events.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/1954332075e9df9b51e1b04eecbb22537308f102/keps/sig-instrumentation/647-apiserver-tracing
```

/sig instrumentation
/kind feature
/priority important-soon

### CODE REVIEW

**Comment by dashpole** [comment]
Ah, I didn't realize k8s.io/utils was the source of truth for that library.  I'll have to make an update there first.

**Comment by dashpole** [comment]
cc @dgrisonnet @logicalhan

**Comment by dims** [comment]
@swagatbora90 if you wanna do a code review!

**Comment by dashpole** [comment]
This is now ready for review.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 1]
```diff
@@ -203,8 +207,11 @@ func (s *store) Create(ctx context.Context, key string, obj, out runtime.Object,
 	if out != nil {
 		putResp := txnResp.Responses[0].GetResponsePut()
 		err = decode(s.codec, s.versioner, data, out, putResp.Header.Revision)
-		trace.Step("decode finished", utiltrace.Field{Key: "len", Value: len(data)}, utiltrace.Field{Key: "err", Value: err})
-		return err
+		if err != nil {
+			span.AddEvent("decode failed", attribute.Int("len", len(data)), attribute.String("err", err.Error()))
+			return err
+		}
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:213]: nice

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 2]
```diff
@@ -462,27 +471,32 @@ func (s *store) GuaranteedUpdate(
...
 		}
 		putResp := txnResp.Responses[0].GetResponsePut()
 
 		err = decode(s.codec, s.versioner, data, destination, putResp.Header.Revision)
-		trace.Step("decode finished", utiltrace.Field{Key: "len", Value: len(data)}, utiltrace.Field{Key: "err", Value: err})
-		return err
+		if err != nil {
+			span.AddEvent("decode failed", attribute.Int("len", len(data)), attribute.String("err", err.Error()))
+			return err
+		}
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:497]: nice

**Review by logicalhan — COMMENTED** [review]
/lgtm

This looks strictly better than it used to be. Thanks for making these changes!

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 3]
```diff
@@ -462,27 +471,32 @@ func (s *store) GuaranteedUpdate(
 			clientv3.OpGet(key),
 		).Commit()
 		metrics.RecordEtcdRequestLatency("update", s.groupResourceString, startTime)
-		trace.Step("Txn call finished", utiltrace.Field{Key: "err", Value: err})
 		if err != nil {
+			span.AddEvent("Txn call failed", attribute.String("err", err.Error()))
 			return err
 		}
-		trace.Step("Transaction committed")
+		span.AddEvent("Txn call succeeded")
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:478]: this phrasing right before checking `if !txnResp.Succeeded` is confusing...

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:478]: Probably `txn call completed` is better

File: staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go [hunk 4]
```diff
@@ -126,15 +128,15 @@ func (b *backend) processEvents(ev ...*auditinternal.Event) error {
 		list.Items = append(list.Items, *e)
 	}
 	return b.w.WithExponentialBackoff(context.Background(), func() rest.Result {
-		trace := utiltrace.New("Call Audit Events webhook",
-			utiltrace.Field{"name", b.name},
-			utiltrace.Field{"event-count", len(list.Items)})
+		ctx, span := tracing.Start(context.TODO(), "Call Audit Events webhook",
```
[the remark this answers is not in the document]
Comment (dashpole) [reply staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go]: It does not.  I thought about using context.TODO() as a way to track that we need to propagate context here, but the utiltrace.New() call is just as good of a way to track that.
Reverted this change.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 3 above)
Comment (dashpole) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Changed to `txn call completed`

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 1 above)
Comment (dashpole) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:213]: thanks!

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 2 above)
Comment (dashpole) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:497]: Thanks!

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

**Pull request merged**